Cython/Build/Cache.py

```diff
@@ -0,0 +1,189 @@
+import sys
+import os
+import hashlib
+import shutil
+import subprocess
+from ..Utils import safe_makedirs, cached_function
+import zipfile
+from .. import __version__
+
+try:
+    import zlib
+    zipfile_compression_mode = zipfile.ZIP_DEFLATED
+except ImportError:
+    zipfile_compression_mode = zipfile.ZIP_STORED
+
+try:
+    import gzip
+    gzip_open = gzip.open
+    gzip_ext = '.gz'
+except ImportError:
+    gzip_open = open
+    gzip_ext = ''
+
+
+join_path = cached_function(os.path.join)
+
+@cached_function
+def file_hash(filename):
+    path = os.path.normpath(filename)
+    prefix = ('%d:%s' % (len(path), path)).encode("UTF-8")
+    m = hashlib.sha1(prefix)
+    with open(path, 'rb') as f:
+        data = f.read(65000)
+        while data:
+            m.update(data)
+            data = f.read(65000)
+    return m.hexdigest()
+
+
+@cached_function
+def get_cython_cache_dir():
+    r"""
+    Return the base directory containing Cython's caches.
+
+    Priority:
+
+    1. CYTHON_CACHE_DIR
+    2. (OS X): ~/Library/Caches/Cython
+       (posix not OS X): XDG_CACHE_HOME/cython if XDG_CACHE_HOME defined
+    3. ~/.cython
+
+    """
+    if 'CYTHON_CACHE_DIR' in os.environ:
+        return os.environ['CYTHON_CACHE_DIR']
+
+    parent = None
+    if os.name == 'posix':
+        if sys.platform == 'darwin':
+            parent = os.path.expanduser('~/Library/Caches')
+        else:
+            # this could fallback on ~/.cache
+            parent = os.environ.get('XDG_CACHE_HOME')
+
+    if parent and os.path.isdir(parent):
+        return os.path.join(parent, 'cython')
+
+    # last fallback: ~/.cython
+    return os.path.expanduser(os.path.join('~', '.cython'))
+
+
+class Cache:
+
+    def __init__(self, cache, cache_size=None):
+        if cache is True:
+            self.enabled = True
+            self.path = os.path.join(get_cython_cache_dir(), 'compiler')
+        elif cache:
+            self.path = cache
+            self.enabled = True
+        else:
+            self.path = None
+            self.enabled = False
+        self.cache_size = cache_size if cache_size is not None else 1024 * 1024 * 100
```

Comment:
The default cache size would be better placed as a global constant (which users can read and control).

Reply:
Fixed in https://github.com/cython/cython/pull/6117.